### PR TITLE
Implement `ProgramDefinition#getAvailablePredicateQuestionDefinitions`

### PR DESCRIPTION
**Pull request by natsid**
### Description
The is part of the larger story of being able to set show/hide predicates on block definitions.

### Checklist
- [x] Created tests which fail without the change (if possible)
- [ ] Extended the README / documentation, if necessary

### Issue(s)
Works toward #322

### CODE REVIEW

File: universal-application-tool-0.0.1/app/services/program/ProgramDefinition.java [hunk 1]
```diff
@@ -191,6 +192,43 @@ public boolean hasEnumerator(long enumeratorId) {
...
+   * Returns a list of the question definitions that may be used to define predicates on the block
+   * definition with the given ID.
+   *
+   * <p>The available question definitions for predicates satisfy BOTH of the following:
+   *
+   * <ul>
+   *   <li>In a block definition that comes sequentially before the block definition with the given
+   *       ID.
+   *   <li>In a block definition that has the same enumerator ID as the block definition with the
+   *       given ID.
```
Comment (bion) [inline universal-application-tool-0.0.1/app/services/program/ProgramDefinition.java:205]: Issue (blocking): a repeated block should be able to specify a non-repeated (top level) question with its predicate.

Comment (natsid) [reply universal-application-tool-0.0.1/app/services/program/ProgramDefinition.java:205]: Ah, will do.

I thought we were doing an initial implementation without this. The complexity won't show up here, but in the predicate evaluation, I believe.

**Review by kendrickt — COMMENTED** [review]
re: enumerators - looks good to me.

File: universal-application-tool-0.0.1/app/services/program/ProgramDefinition.java [hunk 2]
```diff
@@ -191,6 +192,43 @@ public boolean hasEnumerator(long enumeratorId) {
...
+
+    Optional<Long> maybeEnumeratorId = getBlockDefinition(blockId).enumeratorId();
+    ImmutableList<BlockDefinition> blockDefinitionsWithSameEnumerator =
+        maybeEnumeratorId.isPresent()
+            ? getBlockDefinitionsForEnumerator(maybeEnumeratorId.get())
+            : getNonRepeatedBlockDefinitions();
+
+    ImmutableList.Builder<QuestionDefinition> resultBuilder = new ImmutableList.Builder<>();
+
+    for (BlockDefinition blockDefinition : blockDefinitionsWithSameEnumerator) {
```
Comment (kendrickt) [inline universal-application-tool-0.0.1/app/services/program/ProgramDefinition.java:219]: (nit): can you add to `#getBlockDefinitionsForEnumerator` docstring that the order is important and what the meaning of the order is?

**Review by carolinedanzi — APPROVED** [review]

File: universal-application-tool-0.0.1/app/services/program/ProgramDefinition.java (hunk 2 above)
Comment (natsid) [reply universal-application-tool-0.0.1/app/services/program/ProgramDefinition.java]: Oh yeah, let me do this.

File: universal-application-tool-0.0.1/app/services/program/ProgramDefinition.java [hunk 3]
```diff
@@ -191,6 +192,64 @@ public boolean hasEnumerator(long enumeratorId) {
...
+              .map(ProgramQuestionDefinition::getQuestionDefinition)
+              .collect(Collectors.toList()));
+    }
+
+    return builder.build();
+  }
+
+  private ImmutableList<BlockDefinition> getAvailablePredicateBlockDefinitions(
+      BlockDefinition blockDefinition) throws ProgramBlockDefinitionNotFoundException {
+
```
Comment (kendrickt) [inline universal-application-tool-0.0.1/app/services/program/ProgramDefinition.java:225]:
```suggestion
```

**Review by kendrickt — APPROVED** [review]

File: universal-application-tool-0.0.1/app/services/program/ProgramDefinition.java (hunk 1 above)
Comment (natsid) [reply universal-application-tool-0.0.1/app/services/program/ProgramDefinition.java]: Done!

**Review by bion — APPROVED** [review]

**Pull request merged**